1. Xây dựng cây quyết định dựa vào chỉ số độ lợi thông tin và dự đoán nhãn
a. Đọc dữ liệu từ tập dữ liệu đánh giá chất lượng rượu vang trắng trên trang UCI

In [43]:
import pandas as pd
import numpy as np
from google.colab import files
uploaded = files.upload()

Saving winequality-red.csv to winequality-red.csv


In [44]:
data = pd.read_csv("winequality-red.csv", sep = ';')
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

b. . Dữ liệu có bao nhiêu thuộc tính? Cột nào là cột nhãn? Giá trị của các nhãn (ghi lại
kết quả và code vào file nộp bài)

In [45]:
data.shape
np.unique(data['quality'])
data['quality'].value_counts()

5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

có 12 thuộc tính và nhãn là quality: 
5   ->  681 |
6   ->  638 |
7   -> 199 |
4   -> 53 |
8   ->  18|
3  -> 10|

c. Với tập dữ liệu wineWhite sử dụng nghi thức K-Fold để phân chia tập dữ liệu huấn
luyện với K=50, sử dụng tham số “Shuffle” để xáo trộn tập dữ liệu trước khi phân
chia. Xác định số lượng phần tử có trong tập test và tập huấn luyện nếu sử dụng
nghi thức đánh giá này.

In [46]:
from sklearn.model_selection import KFold
kf = KFold(n_splits=50, shuffle = True)


In [47]:
X = data.iloc[:,:11]
y = data.iloc[:,11]

In [48]:
for train_index, test_index in kf.split(X):
  print("Train:", train_index, "Test:", test_index)
  X_train, X_test = X.iloc[train_index], X.iloc[test_index]
  y_train, y_test = y.iloc[train_index], y.iloc[test_index]
  print("X test", X_test)
  print("==============")
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

Streaming output truncated to the last 5000 lines.
20                   29.0                  60.0  0.99680  3.39       0.53   
98                    5.0                  10.0  0.99670  3.39       0.56   
111                  11.0                 108.0  0.99640  3.15       0.66   
238                  14.0                  38.0  0.99675  3.37       0.58   
269                   4.0                  23.0  0.99960  3.28       0.97   
277                   4.0                  23.0  0.99960  3.28       0.97   
278                   5.0                  13.0  0.99760  3.23       0.82   
281                   5.0                  10.0  0.99720  3.25       1.08   
336                   6.0                  16.0  0.99480  3.35       0.70   
446                  31.0                  72.0  0.99960  3.10       0.73   
480                   6.0                  23.0  1.00260  3.12       0.66   
497                  43.0                 113.0  0.99660  3.32       0.79   
656                   8.0

In [49]:
print(X_train.shape)
X_test.shape

(1568, 11)


(31, 11)

ta có 1568 tập huấn luyện và 31 tập kiểm tra

d. Xây dựng mô hình cây quyết định dựa trên tập dữ liệu học tạo ra ở bước c.

In [50]:
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(criterion="gini", random_state=100, max_depth = 5,min_samples_leaf=5)
clf.fit(X_train, y_train)


DecisionTreeClassifier(max_depth=5, min_samples_leaf=5, random_state=100)

e. Đánh giá độ chính xác cho từng phân lớp dựa vào giá trị dự đoán của câu 3 cho mỗi
lần lặp. Chép lại kết quả độ chính xác cho từng phân lớp của lần lặp cuối nộp lại
(có thể đưa vào comment trong file code).

In [51]:
y_pred = clf.predict(X_test)

In [52]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)*100

70.96774193548387

độ chính xác là:45.16129032258064

```
# This is formatted as code
```



f. Tính độ chính xác tổng thể cho mỗi lần lặp và độ chính xác tổng thể của trung bình
50 lần lặp

In [53]:
sum = 0
for i in range(10, 61):
  kf = KFold(n_splits=i, shuffle = True)
  X1 = data.iloc[:,:11]
  y1 = data.iloc[:,11] 
  for train_index, test_index in kf.split(X1):
    X_train1, X_test1 = X1.iloc[train_index], X1.iloc[test_index]
    y_train1, y_test1 = y1.iloc[train_index], y1.iloc[test_index]
  clf = DecisionTreeClassifier(criterion="gini", random_state=100, max_depth = 5,min_samples_leaf=5)
  clf.fit(X_train1, y_train1)
  y_pred1 = clf.predict(X_test1)
  sum += accuracy_score(y_test1, y_pred1)*100
sum / 50

57.865349454661754

Trung bình độ chính xác trong 50 lần lặp là: 59.1

g. Sử dụng giải thuật KNN, Bayes thơ ngây ở buổi thực hành số 2 để so sánh hiệu quả
phân lớp của giải thuật cây quyết định với nghi thức đánh giá k-fold với K=60

- Thuật toán knn

In [54]:
from sklearn.model_selection import train_test_split
Xtrain,Xtest, ytrain, ytest = train_test_split(X,y,random_state=42, test_size = 1/3.0)


In [55]:
from sklearn.neighbors import KNeighborsClassifier
mohinh_knn = KNeighborsClassifier(n_neighbors = 5)
mohinh_knn.fit(Xtrain,ytrain)
y_pred_knn = mohinh_knn.predict(Xtest)
Accuracy_knn = accuracy_score(y_pred_knn,ytest)*100
Accuracy_knn


47.467166979362105

Độ chính xác với thuât toán knn là: 47.5

- Thuật toán Bayes thơ ngây

In [56]:
from sklearn.naive_bayes import GaussianNB
GNB = GaussianNB()
GNB.fit(Xtrain,ytrain)
y_predGNB = GNB.predict(Xtest)
Accuracy_GNB = accuracy_score(y_predGNB,ytest) * 100
Accuracy_GNB

54.78424015009381

Độ chính xác của thuật toán Bayes: 54.8

- Giải thuật cây quyết định với nghi thức đánh giá k-fold với K=60

In [57]:
kf = KFold(n_splits=60)
for train_index, test_index in kf.split(X1):
    X_train1, X_test1 = X1.iloc[train_index], X1.iloc[test_index]
    y_train1, y_test1 = y1.iloc[train_index], y1.iloc[test_index]
clf = DecisionTreeClassifier(criterion="gini", random_state=100, max_depth = 5,min_samples_leaf=5)
clf.fit(X_train1, y_train1)
y_pred1 = clf.predict(X_test1)
Accuracy_DCT = accuracy_score(y_pred1, y_test1)*100
Accuracy_DCT

65.38461538461539

Độ chính xác của thuật toán cây quyết định là: 65.4
Vậy độ chính xác của Cây quyết định > Bayes > knn

2. Cho tập dữ liệu gồm 5 phần tử như bảng bên dưới,


In [58]:
from google.colab import files
uploaded = files.upload()

Saving gioitinh.csv to gioitinh (5).csv


In [59]:
dt = pd.read_csv('gioitinh.csv',sep=' ')
X = dt.iloc[:,:3]
y = dt.iloc[:,3]



In [60]:
kf = KFold(n_splits=2)
for train_index, test_index in kf.split(X):
    X_train1, X_test1 = X.iloc[train_index], X.iloc[test_index]
    y_train1, y_test1 = y.iloc[train_index], y.iloc[test_index]
clf = DecisionTreeClassifier(criterion="gini", random_state=100, max_depth = 5,min_samples_leaf=5)
clf.fit(X_train1, y_train1)


DecisionTreeClassifier(max_depth=5, min_samples_leaf=5, random_state=100)

Dự báo phần tử mới tới có thông tin chiều cao=135, độ dài mái tóc = 39 và giọng nói
có giá trị là 1 thì người này là 0 hay 1?

In [67]:
pt = pd.DataFrame({
    "chieucao":[135],
    "toc":[39],
    "giongnoi":[1]
})
clf.predict(pt)

array([1])

Vậy là nữa